### PR TITLE
Text Embedding: Truncation allow or fail with optional parameter

### PR DESCRIPTION
**Pull request by markstur**
The desired functionality here is that a request which would be truncated (to work with a model max_seq_len) should instead return an error message.  The reasoning is that the client would want to know, instead of just getting results without realizing the impact of truncation.

In order to consider the token truncation (not string truncation), this requires tokenization up front and avoiding a warning about parallelism on the first attempt.

To be consistent with the user experience with other models, the default will throw an error with the models max_seq_len is exceeded (this is not the sentence-transformers default behavior so it is implemented here).  In addition, here we add the optional truncate_input_tokens parameter.  This is consistent with other services, and in addition to implementing the desired default behavior, this implements a passed-in truncation, or using sentence-transformers automatically truncate with -1 (and less overhead that way).

            truncate_input_tokens: int
                Truncation length for input tokens.
                If less than zero, this is disabled (returns texts without processing).
                If zero or greater than the model's maximum, then this is a test
                to see if truncation is needed. If needed, an exception is thrown.
                Otherwise, we take this usable truncation limit to truncate the tokens and then
                decode them to return truncated strings that can be used with this model.

### CODE REVIEW

File: caikit_nlp/modules/text_embedding/embedding.py [hunk 1]
```diff
@@ -46,6 +46,11 @@
 logger = alog.use_channel("TXT_EMB")
 error = error_handler.get(logger)
 
+# Avoid a warning by explicitly setting TOKENIZERS_PARALLELISM if not already set.
```
Comment (markstur) [inline caikit_nlp/modules/text_embedding/embedding.py:49]: More explanation...

Background detail:   This came up in a scenario where lazy loading is used.   So the model is always loaded on first request, and with this truncation code the tokenizer is always immediately used.  So this error message is consistent on first use.

```
2023-11-28T18:07:19.695712 [MODEL-SIZER :INFO:8166760128] <RUN62161564I> No configured model size multiplier found for model type 'standalone-model' for model 'example/model22'. Using default multiplier '10.000000'
2023-11-28T18:07:19.698317 [sentence_tra:INFO:11295420416] Load pretrained SentenceTransformer: /Users/markstur/caikit-nlp/models/example/model22/artifacts
2023-11-28T18:07:20.679813 [sentence_tra:INFO:11295420416] Use pytorch device: cpu
2023-11-28T18:07:20.679940 [MODEL-LOADER:INFO:11295420416] <RUN89713784I> Singleton cache: '{}'
huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
        - Avoid using `tokenizers` before the fork if possible
        - Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
```

So, the 2nd option is implemented here by setting the env if not already set.  The nice thing about this is that any deployment can set the environment if they don't want this default.

There is a simple alternative that might be less ugly(?).  If I just add a single call to model.encode("foo") at the end of load() that seems to avoid using tokenizers too early, because essentially the model (and tokenizer) is primed and ready.  This seems like the less straight-forward way to address this specific problem, but it is a nice single line of code and a quick model test to make sure the model is "ready".  Unfortunately with lazy load this warmup is on the first request anyway, but otherwise I might like some amount of warmup.

**Review by evaline-ju — COMMENTED** [review]
a few questions and some test requests

File: caikit_nlp/modules/text_embedding/embedding.py [hunk 2]
```diff
@@ -120,28 +120,131 @@ def load(cls, model_path: str, *args, **kwargs) -> "EmbeddingModule":
...
+            okay_to_truncate = False
+            max_length = max_tokens
+            ret = texts  # will not alter texts when truncation is not allowed
+
+        for text in texts:
+            tokenized = self.model.tokenizer(
+                text,
+                return_attention_mask=False,
+                return_token_type_ids=False,
+                return_overflowing_tokens=True,
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/embedding.py:171]: I don't actually use the tokens, but this setting gives me the overflow which indicates that there were extra tokens. This allows me to (by default) detect overflow, count up the total, and raise the error.  This is the primary goal... to make sentence-transformers fail like other models instead of silently truncating when the model max is exceeded.  The error message is expected to provide the token count because it isn't exactly character count.  

Since I also allow a truncation setting <= model_max (when the client sends that parameter). This setting also gives me the input_ids that I'll use for that allowed truncation.

The best alternative I considered was to pad the max and then do the truncation myself, but I think the current way provides a better way to truncate and a better way to detect overflow and report the total length.

File: caikit_nlp/modules/text_embedding/embedding.py [hunk 3]
```diff
@@ -120,28 +120,131 @@ def load(cls, model_path: str, *args, **kwargs) -> "EmbeddingModule":
...
+
+        # Do truncation if given a usable truncation value, else test for need to truncation
+        if 0 < truncate_input_tokens <= max_tokens:
+            okay_to_truncate = True
+            max_length = truncate_input_tokens
+            ret = []  # will return truncated texts
+        else:
+            okay_to_truncate = False
+            max_length = max_tokens
+            ret = texts  # will not alter texts when truncation is not allowed
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/embedding.py:164]: In this case, we don't allow truncation but raise an exception which is returned.  So the texts are returned as-is only if we get through that.  Have to loop through to determine if the model would in fact do truncation then do the `elif was_truncated` ... `error.log_raise` (below).  

Unfortunately sentence-transformers doesn't just let me pass some parameter to behave this way :(. So truncation would happen with sentence-transformers, but the feature requested here is to raise an exception instead (to be consistent with other services/user experiences).

Since you mentioned "last error", I should point out that this raises an exception as soon as unwanted truncation is detected. There isn't any attempt to look for more errors right now - just fail fast.  The unit test is mostly checking that the exception is thrown when expected.

```
    model.run_embedding(text=ok)
    with pytest.raises(ValueError):
        model.run_embedding(text=too_long)
```

File: caikit_nlp/modules/text_embedding/embedding.py [hunk 4]
```diff
@@ -120,28 +120,131 @@ def load(cls, model_path: str, *args, **kwargs) -> "EmbeddingModule":
...
+            )
+
+            lengths = tokenized["length"]
+            was_truncated = len(lengths) > 1  # multiple lengths when truncated
+
+            if okay_to_truncate and was_truncated:
+                # decode the truncated input tokens back to text to be returned
+                ret.append(
+                    self.model.tokenizer.decode(
+                        tokenized.input_ids[0], skip_special_tokens=True
```
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/embedding.py:184]: Good catch.  It's better not to skip them.  You make a good point about truncated vs not-truncated inconsistency.  I can also see how we could compare 2 inputs with special tokens (unfortunately?).  In that case, the match results are better if we error on the side of keeping them than if we error on the side of skipping the wrong ones.

This doesn't seem perfect, but I'll change that to False.  That's probably the best we can do here.  The caller can send -1 to let the model truncate if this truncate w/ decode has to be avoided.

File: tests/modules/text_embedding/test_embedding.py [hunk 5]
```diff
@@ -378,3 +378,47 @@ def test_run_sentence_similarities():
...
+    """These endpoints raise an error if truncation would happen."""
+
+    model = BOOTSTRAPPED_MODEL
+    model_max = model.model.max_seq_length
+
+    ok = "x " * (model_max - 2)  # Subtract 2 for begin/end tokens
+    too_long = "x " * (model_max - 1)  # This will go over
+
+    # embedding(s)
+    model.run_embedding(text=ok)
```
[the remark this answers is not in the document]
Comment (markstur) [reply tests/modules/text_embedding/test_embedding.py:393]: added more tests and split them more into happy/unhappy, params, etc...

File: tests/modules/text_embedding/test_embedding.py [hunk 6]
```diff
@@ -378,3 +378,47 @@ def test_run_sentence_similarities():
         assert len(scores) == len(SENTENCES)
         for score in scores:
             assert isinstance(score, float)
+
+
+def test_no_truncation():
```
[the remark this answers is not in the document]
Comment (markstur) [reply tests/modules/text_embedding/test_embedding.py:383]: fixed

File: caikit_nlp/modules/text_embedding/embedding.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/embedding.py:184]: fixed

**Comment by markstur** [comment]
rebased on main

File: caikit_nlp/modules/text_embedding/embedding.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (markstur) [reply caikit_nlp/modules/text_embedding/embedding.py:258]: added more tests

[the remark this answers is not in the document]
Comment (evaline-ju) [reply caikit_nlp/modules/text_embedding/embedding.py:258]: I see. My main concern here was if more "work" was being done than necessary

**Review by evaline-ju — APPROVED** [review]
LGTM! Thanks for all the explanations and tests
